In [22]:
import pandas as pd
from sqlalchemy import create_engine
import warnings

postgres_user = ''
postgres_pw = ''
postgres_host = ''
postgres_port = '5432'
postgres_db = 'weatherinszeged'

In [21]:

engine2 = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df = pd.read_sql_query('SELECT * FROM weatherinszeged',con=engine2)
engine2.dispose()

##  Weather model

In [3]:
df.head()

date        summary preciptype  temperature  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy       rain     9.472222   
1 2006-03-31 23:00:00+00:00  Partly Cloudy       rain     9.355556   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy       rain     9.377778   
3 2006-04-01 01:00:00+00:00  Partly Cloudy       rain     8.288889   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy       rain     8.755556   

   apparenttemperature  humidity  windspeed  windbearing  visibility  \
0             7.388889      0.89    14.1197        251.0     15.8263   
1             7.227778      0.86    14.2646        259.0     15.8263   
2             9.377778      0.89     3.9284        204.0     14.9569   
3             5.944444      0.83    14.1036        269.0     15.8263   
4             6.977778      0.83    11.0446        259.0     15.8263   

   loudcover  pressure                       dailysummary  
0        0.0   1015.13  Partly cloudy throughout the day.  
1        0.0   1015.63  Partly cloudy throughout the day.  
2        0.0   1015.94  Partly cloudy throughout the day.  
3        0.0   1016.41  Partly cloudy throughout the day.  
4        0.0   1016.51  Partly cloudy throughout the day.

In [6]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import statsmodels.api as sm
from sqlalchemy import create_engine
# `Y` is the target variable is the difference between the apparenttemperature and the temperature
Y = df['temperature'] -df['apparenttemperature']

# `X` is the feature set
X = df[['humidity','windspeed']]

# Add a constant to the model because it's best practice
# to do so every time!
X = sm.add_constant(X)

# Fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# Print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Wed, 18 Sep 2024   Prob (F-statistic):               0.00
Time:                        15:29:40   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.4381      0.021   -115.948      0.0

In [7]:
df["humidity_windspeed"] = df.humidity * df.windspeed

In [8]:
# `X` is the feature set
X = df[['humidity','windspeed','humidity_windspeed']]

# Add a constant to the model because it's best practice
# to do so every time!
X = sm.add_constant(X)

# Fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# Print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Wed, 18 Sep 2024   Prob (F-statistic):               0.00
Time:                        15:31:18   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -0.0839      0

In [9]:
#--- add visibility
#--- add visibility

# `X` is the feature set
X = df[['humidity','windspeed','visibility']]

# Add a constant to the model because it's best practice
# to do so every time!
X = sm.add_constant(X)

# Fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# Print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Wed, 18 Sep 2024   Prob (F-statistic):               0.00
Time:                        15:32:46   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.5756      0.028    -56.605      0.0

In [10]:
# `X` is the feature set
X = df[['humidity','windspeed','humidity_windspeed','visibility']]

# Add a constant to the model because it's best practice
# to do so every time!
X = sm.add_constant(X)

# Fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# Print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.364
Model:                            OLS   Adj. R-squared:                  0.363
Method:                 Least Squares   F-statistic:                 1.377e+04
Date:                Wed, 18 Sep 2024   Prob (F-statistic):               0.00
Time:                        15:33:05   Log-Likelihood:            -1.6504e+05
No. Observations:               96453   AIC:                         3.301e+05
Df Residuals:                   96448   BIC:                         3.301e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                  1.1006      0

### summary
from above models and r square values , the model which has high score is the one including humidity, windspeed and visibilty in addition the interaction variable between wind spedd and humidity

## House price model

In [12]:
postgres_db='houseprices'
engine2 = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

houseprices_df = pd.read_sql_query('select * from houseprices',con=engine2)
engine2.dispose()

In [13]:
# select features  year build, sqfeet , lot area , full bath ,halfbath, overallcond, live area
fea_df = houseprices_df[['yearbuilt','lotarea','fullbath','halfbath','grlivarea','overallcond','saleprice']]

In [14]:
# bath and condition categorical
bath_df = pd.get_dummies(fea_df['fullbath'],dtype=int).rename(columns=lambda x:'fb_' +str(x))
new_df = fea_df.drop(['fullbath'], axis =1).merge(bath_df,left_index=True, right_index=True)

In [15]:
bath_df = pd.get_dummies(fea_df['halfbath'],dtype=int).rename(columns=lambda x:'hb_' +str(x))
new2_df = new_df.drop(['halfbath'], axis =1).merge(bath_df,left_index=True, right_index=True)

In [16]:
new2_df.head()

yearbuilt  lotarea  grlivarea  overallcond  saleprice  fb_0  fb_1  fb_2  \
0       2003     8450       1710            5     208500     0     0     1   
1       1976     9600       1262            8     181500     0     0     1   
2       2001    11250       1786            5     223500     0     0     1   
3       1915     9550       1717            5     140000     0     1     0   
4       2000    14260       2198            5     250000     0     0     1   

   fb_3  hb_0  hb_1  hb_2  
0     0     0     1     0  
1     0     1     0     0  
2     0     0     1     0  
3     0     1     0     0  
4     0     0     1     0

In [17]:
# `Y` is the target variable
Y = new2_df['saleprice']
# `X` is the feature set which includes the
# `is_male` and `is_smoker` variables
X = new2_df.drop(['saleprice'], axis =1)

In [18]:
import statsmodels.api as sm

# You need to manually add a constant
# in statsmodels' sm
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.696
Model:                            OLS   Adj. R-squared:                  0.694
Method:                 Least Squares   F-statistic:                     369.4
Date:                Wed, 18 Sep 2024   Prob (F-statistic):               0.00
Time:                        15:40:52   Log-Likelihood:                -17674.
No. Observations:                1460   AIC:                         3.537e+04
Df Residuals:                    1450   BIC:                         3.542e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.682e+06   6.31e+04    -26.655      0.000   -1.81e+06   -1.56e+06
yearbuilt    1330.9623     48.777     27.287      0.000    1235.282    1426.642
lotarea         0.6810      0.121      5.647      0.000       0.444       0.918
grlivarea     101.4269      3.387     29.948      0.000      94.783     108.070
overallcond  9907.3360   1122.709      8.824      0.000    7705.029    1.21e+04
fb_0        -3.922e+05   1.92e+04    -20.466      0.000    -4.3e+05   -3.55e+05
fb_1          -4.3e+05   1.53e+04    -28.071      0.000    -4.6e+05      -4e+05
fb_2        -4.445e+05   1.69e+04    -26.339      0.000   -4.78e+05   -4.11e+05
fb_3        -4.149e+05   1.88e+04    -22.026      0.000   -4.52e+05   -3.78e+05
hb_0        -5.345e+05   2.09e+04    -25.544      0.000   -5.76e+05   -4.93e+05
hb_1        -5.489e+05   2.18e+04    -25.234      0.000   -5.92e+05   -5.06e+05
hb_2        -5.981e+05   2.32e+04    -25.772      0.000   -6.44e+05   -5.53e+05
==============================================================================
Omnibus:                      390.195   Durbin-Watson:                   2.009
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            27402.846
Skew:                          -0.190   Prob(JB):                         0.00
Kurtosis:                      24.221   Cond. No.                     7.69e+19
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 5.27e-29. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

This model's F-statistic is 369.4, and the associated p-value is very close to zero. This means that your features add some information to the reduced model, and your model is useful in explaining house prices.
AIC/BIC IS QUITE BIG , WE MAY NEED TO SELECT MORE FEATURES TO GET BETTER MODEL 
R SQARE IS .696 , MODEL IS SOME WHAT EXPLAINING THE HOUSE PRICE , ADDING OTHER VARIABLES MIGHT IMPROVE PERFORMANCE